components-java-sdk/src/main/java/io/dapr/components/wrappers/RequestStreamToFluxAdaptor.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.components.wrappers;
+
+import io.dapr.components.domain.bindings.ReadRequest;
+import io.dapr.components.domain.pubsub.PublishRequest;
+import io.grpc.stub.StreamObserver;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Sinks;
+
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import static reactor.core.publisher.Sinks.EmitFailureHandler.FAIL_FAST;
+
+/**
+ * Creates a StreamObserver that adapts and feeds its received requests to a Flux.
+ *
+ * <p>It also does peaks the very first response and stores it aside becasue both
+ * {@link io.dapr.components.domain.bindings.InputBinding#read(ReadRequest, Flux)} and
+ * {@link io.dapr.components.domain.pubsub.PubSub#publish(PublishRequest)} have
+ * this weird thing in that the very first message in the request/input stream is special.
+ * </p>
+ *
+ * <p>>gRPC is wierd in that a bi-directional stream receive its input requests by
+ * means of a {@link StreamObserver} its RPC message handler returns. Think about it:
+ * we return responses on a StreamObserver we receive as argument but we receive requests in a
+ * StreamObserver we return as result. Mind bogging, right? </p>
+ *
+ * <p>Well, the process of adapting this "returned" {@link StreamObserver} to what
+ * will turn into an input {@link Flux} by means of an intermediary {@link Sinks.Many} is
+ * quite repetitive. This adaptor class simplifies this process. Rejoice.</p>
+ *
+ * @param <StreamT> The input stream type as per the gRPC interface.
+ */
+public final class RequestStreamToFluxAdaptor<StreamT> {
+  private final StreamObserver<StreamT> requestStreamObserver;
+
+  private boolean isThisTheFirstItem = true;
+
+  /**
+   * Constructor.
+   *
+   * @param onFirstCallback the callback we invoke upon receiving the very first request.
+   *                       Its first argument is the very first request received and the second
+   *                       the remaining requests as a flux.
+   */
+  public RequestStreamToFluxAdaptor(BiConsumer<StreamT, Flux<StreamT>> onFirstCallback) {
+    final Sinks.Many<StreamT> sink = Sinks.many().unicast().onBackpressureBuffer();
+    requestStreamObserver = adaptRequestStreamToSink(sink, onFirstCallback);
+  }
+
+  // TODO(tmacam) perhaps this whole thing could be replaced by a single static method that receives the bi-func
```

Comment:
I wonder if it'd be simpler to first adapt a `StreamObserver` to a `Flux`, then do the extraction of the first element and remainders using the Flux operators.

Reply:
Oh! I thought so. But I spent a whole day trying to do that and failed miserably . One of the problems is: this particular Sink I create limits access to the stream to a single subscriber and it doesn't keep past items in memory. Which seems like the appropriate thing to do, honestly. But once you try to, say, fetch "the head" and "the tail" of a stream, essentially, you are subscribing twice to this stream and if fails in a rather ugly way. 

So things that work when you create tests with `Flux.just(<array here>)` fail miserably (as in dead lock) when you try to create two fluxes from it: a head and a tail one. The only way I managed to get it to work was using an `flux.cache()`, that creates an ConnectableFlux (think "materialized view" if this was a DB) that caches **everything**.  

I can try to play with that again because it would be WAY more elegant.